<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>SnipSnip: GWAS for Poorly Tagged Data using Multiple SNPs</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<link rel="stylesheet" type="text/css" href="snipsnipstyles.css" />
<link rel="shortcut icon" href="favicon.ico" />
</head>
<body>
<!-- Begin Wrapper -->
<div id="wrapper">
<table width="100%" border="0" cellpadding="0" cellspacing="0"><tr><td>
<!-- Begin Header -->
<div id="header">
<table><tr>
<td>
<img width="200" src="snipsniplogo.png" alt="SnipSnip: GWAS for Poorly Tagged Data using Multiple SNPs"/>
</td>
<td>
<p style="margin-left: 1em;">SnipSnip: GWAS for Poorly Tagged Data using Multiple SNPs</p>
</td></tr>
</table>
</div>
<!-- End Header -->
<!-- Begin Menu Navigation -->
<ul id="menunav">
<li><a href="index.html">Home</a></li>
<li><a href="download.html">Download</a></li>
<li><a href="contact.html">Contact</a></li>
</ul>
<!-- End Menu Navigation -->

</td></tr>
<tr><td>
<table border="0" cellpadding="0" cellspacing="0"><tr>
<td valign="top">
<!-- Begin Left Column -->
<div id="leftcolumn">
<!-- Begin Contents Navigation -->
<div id="navcontainer">
<ul>
<li><a href="introduction.html">1 Introduction</a>
<ul>
<li><a href="introduction.html#information">1.1 Program information and citation</a></li>
</ul>
</li>
<li><a href="installation.html">2 Installation</a>
</li>
<li><a href="using.html">3 Using SnipSnip</a>
<ul>
<li><a href="using.html#basic-usage">3.1 Basic Usage</a></li>
<li><a href="using.html#options">3.2 Options</a></li>
<li><a href="using.html#parameterfile">3.3 Parameter file</a></li>
<li><a href="using.html#input">3.4 Input</a></li>
<li><a href="using.html#output">3.5 Output</a></li>
</ul>
</li>
<li><a href="snipsnip-gwas.html">4 SnipSnip GWAS</a>
<ul>
<li><a href="snipsnip-gwas.html#anchor-snp">4.1 Anchor SNP</a></li>
<li><a href="snipsnip-gwas.html#partner-snp">4.2 Partner SNP</a></li>
<li><a href="snipsnip-gwas.html#ai-test">4.3 AI Test</a></li>
</ul>
</li>
<li><a href="metrics.html">5 Partner SNP Metrics</a>
</li>
<li><a href="windowsize.html">6 SNP Window Size</a>
</li>
<li><a href="linear.html">7 Linear Regression</a>
</li>
<li><a href="covariates.html">8 Covariates</a>
</li>
<li><a href="example.html">9 SnipSnip Examples</a>
<ul>
<li><a href="example.html#ex-basic">9.1 Basic example</a></li>
<li><a href="example.html#X-chromosome">9.2 X chromosome</a></li>
</ul>
</li>
<li><a href="references.html">References</a></li>
</ul>
</div>
<!-- End Contents Navigation -->
</div>
<!-- End Left Column -->
</td>
<td valign="top">
<!-- Begin Right Column -->
<div id="rightcolumn">
<h1>9 SnipSnip Examples</h1>
<h2><a name="ex-basic">9.1 Basic example</a></h2>
<p>
Using the example data given with the SnipSnip download, perform a basic SnipSnip analysis as follows: </p>
<pre>
./snipsnip -o results-ExampleData.dat exampleData.bed
</pre>
<p>
In R type: </p>
<pre>
exampleData&lt;-read.table("results-ExampleData.dat", header=T)

plot(exampleData$BP/10^6, -log10(exampleData$P), main="SnipSnip Test",
 xlab=expression(bp~position~(Mb)), ylab=expression(-log[10](p-value)), ylim=c(0,15)) 
abline(h=8, lty=2)
</pre>
<p>
This will produce the following plot: </p>
<div id="fig"><a href="exampleSS.png"><img src="exampleSS.png" border="0" class="figimg" width="300"></a><br />
<br />
Figure 3. Manhattan plot of AI test results.
</div>
<p>
To also perform standard logistic regression use the <tt>-lr</tt> option as follows: </p>
<pre>
./snipsnip -lr -o results-ExampleData-LR.dat exampleData.bed
</pre>
<p>
To plot the standard logistic regression test results type: </p>
<pre>
exampleDataLR&lt;-read.table("results-ExampleData-LR.dat", header=T)

plot(exampleDataLR$BP/10^6, -log10(exampleDataLR$P_LR), main="Standard Logistic Regression",
 xlab=expression(bp~position~(Mb)), ylab=expression(-log[10](p-value)), ylim=c(0,15)) 
abline(h=8, lty=2)

</pre>
<p>
This will produce the following plot: </p>
<div id="fig"><a href="exampleSS_LR.png"><img src="exampleSS_LR.png" border="0" class="figimg" width="300"></a><br />
<br />
Figure 4. Manhattan plot of standard logistic regression test results.
</div>
<h2><a name="X-chromosome">9.2 X chromosome</a></h2>
<p>
SnipSnip does not do anything special with the analysis of SNPs on the X chromosome - the SNPs will be treated as if they were autosomal. (So females will end up with genotypes coded 0,1,2 and males will end up with genotypes coded 0,2). This has the potential to create false positives if you are dealing with a disease that has different prevalance in males and females; for this reason it is recommended to always include gender as a covariate if you are analysing X-chromosomal SNPs. See <a href="covariates.html">section 8</a> for details on handling covariates. </p>
<p>
If you have no other covariates to consider then it is possible to use the <tt>.fam</tt> file as the covariate file by choosing the sex column as the covariate as follows: </p>
<pre>
./snipsnip -covar chromosome23.fam -covar-number 3 -o myresults-chr23.dat chromosome23.bed
</pre>
<br />
<div class="prevnext"><span class="left"><a href="covariates.html">&lt;-prev</a>
</span></div><!-- End Right Column -->
</div>
</td>
</tr></table>
<!-- End Wrapper -->
</td></tr></table>
</div>
 <!-- Start of StatCounter Code for Default Guide --> <script type="text/javascript"> var sc_project=8908722; var sc_invisible=1; var sc_security="87584528"; var scJsHost = (("https:" == document.location.protocol) ? "https://secure." : "http://www."); document.write("<sc"+"ript type='text/javascript' src='" + scJsHost+ "statcounter.com/counter/counter.js'></"+"script>"); </script> <noscript><div class="statcounter"><a title="web analytics" href="http://statcounter.com/" target="_blank"><img class="statcounter" src="http://c.statcounter.com/8908722/0/87584528/1/" alt="web analytics"></a></div></noscript> <!-- End of StatCounter Code for Default Guide</body>
</html>
